README.md
# 🛠️ Beyond Repair (Master 0.0.2)
[![Contributors][contributors-shield]][contributors-url]
[![Forks][forks-shield]][forks-url]
[![Stargazers][stars-shield]][stars-url]
[![Issues][issues-shield]][issues-url]
[![Lua][lua-shield]][lua-url]<br/>
Beyond-Repair will focus on making it easy to semi/fully automate tasks by having ready to use functions & examples.
As I've released the first stable version of the addon my focus will shift making a stable version of AHK
![Frame screenie](./Misc/frames.png?raw=true "Frames screenshot")

## 🔥 Current features
* Works with 3.x.x 
* Using WowArc3 Lib

## 🚀 Future features
* Bypass blocked lua
* AHK program
* Solo/Party/Follower mode
* Compatible /w mainstream versions

## ✨ Change log
0.0.**2** - Creating AHK script<br/>
0.0.**1** - Updated repo; Code upgrade<br/>

## ⭐️ Useful info
* BTP [Github](https://github.com/godsflaw/btp)/[Launchpad](https://launchpad.net/btp)/[Slideshow](./Misc/defconslideshow.pdf)<br/>
* [WowArc Lib](https://www.wowace.com/projects/ace3)<br/>
* [Lua Doc](https://www.lua.org/pil/contents.html)<br/>

## 🤝 Contributing
Contributions, issues and feature requests are welcome.<br/>
Feel free to check [issues page](https://github.com/nizonrox/Beyond-Repair/issues) if you want to contribute.<br/>
[Check the contributing guide](./CONTRIBUTING.md).<br/>

## 📝 License
[Apache-2.0 License](./LICENSE)

[contributors-shield]: https://img.shields.io/github/contributors/nizonrox/Beyond-Repair.svg?style=for-the-badge
[contributors-url]: https://github.com/nizonrox/Beyond-Repair/graphs/contributors
[forks-shield]: https://img.shields.io/github/forks/nizonrox/Beyond-Repair.svg?style=for-the-badge
[forks-url]: https://github.com/nizonrox/Beyond-Repair/network/members
[stars-shield]: https://img.shields.io/github/stars/nizonrox/Beyond-Repair.svg?style=for-the-badge
[stars-url]: https://github.com/nizonrox/Beyond-Repair/stargazers
[issues-shield]: https://img.shields.io/github/issues/nizonrox/Beyond-Repair.svg?style=for-the-badge
[issues-url]: https://github.com/nizonrox/Beyond-Repair/issues
[lua-shield]: https://img.shields.io/badge/Lua-2C2D72?style=for-the-badge&logo=lua&logoColor=white
[lua-url]: https://www.lua.org/cgi-bin/demo


changelog.txt
0.0.1 - Stable
- Fixed debugging / Added on the fly debugging
- Intergrated WoWArc lib
    -Added commands
    -Changed from xml onload to Arc3 OnInitialize
- Changed function naming scheme to avoid function name overlapping
- Added comments to a lot of functions
- Improved frame creation/recoloring
- Removing clutter


beyond/beyond_lib.lua
-- Globals
cur_xpos_g = 0
cur_ypos_g = 0
fr_width_g = 75
fr_height_g = 20
beyond_f_array = { "IT", "IA", "IPT", "CT", "CA", "CPT", "AT", "AA", "APT", "PT", "MA", "PA", "MP", "PPT" }
beyond_c_array = { "000011", "000022", "000033", "000044", "000055", "000066", "000077", "000088", "000099", "0000AA", "0000BB", "0000BB", "0000CC", "0000DD", "0000EE" }
beyond_debug = 0

-- Debug message for trouble shooting; With On/Off option via variable
function beyond_lib_debug(msg)
    if beyond_debug=1 then
        DEFAULT_CHAT_FRAME:AddMessage(msg)
        UIErrorsFrame:AddMessage(msg, 1.0, 1.0, 0, 1, 10)
    Else
        return
    end
end

-- Calculate where the next frame should be (Used in btp_frame_onload)
function btp_next_cord()
    local curx
    local cury
    curx = cur_xpos_g
    cury = cur_ypos_g

    cur_xpos_g = cur_xpos_g + fr_width_g
    beyond_lib_debug("xpos: " .. curx .. " ypos: " .. cury)
    return curx, cury
end

-- Loops over each frame and initializes them by changing their hex
function beyond_lib_init()
    for i,v in pairs(beyond_f_array) do
        curx, cury = btp_next_cord()
        beyond_lib_debug("Change: " .. beyond_f_array[i] .. " Hex: " .. beyond_c_array[i])
        btp_frame_set_color_hex(beyond_f_array[i], beyond_c_array[i])
    end
end

-- Define slash commands/ Creation of all frames
function btp_frame_onload(self)
    local curx
    local cury
    local curw
    local curh
    curw = fr_width_g
    curh = fr_height_g

    beyond_lib_debug("btp_frame: OnLoad")
    SLASH_BTP_FRAME1 = "/btp_frame"
    SLASH_BTP_INIT1 = "/btp_finit"
    SLASH_BTP_DBG1 = "/btp_dbg"

    SlashCmdList["BTP_FRAME"] = function(cmdstr)
        -- btp_frame_debug("cmdstr: " .. cmdstr)
        btp_frame_runcmd(cmdstr)
    end

    SlashCmdList["BTP_INIT"] = beyond_lib_init

    SlashCmdList["BTP_DBG"] = function(cmdstr)
        -- btp_frame_debug(cmdstr);
    end
    
    --Loop creation of bars
    for i,v in pairs(beyond_f_array) do
        curx, cury = btp_next_cord()
        btp_frame_create(beyond_f_array[i], curx, cury, curw, curh)
    end
end

-- Hide/Show function
function btp_frame_toggle(fname)
    local full_name = "btp_frame_" .. fname
    local frame = getglobal(full_name)
    --btp_frame_debug(full_name)
    if (frame) then
        if (frame:IsVisible()) then
            frame:Hide()
        else
            frame:Show()
        end
    end
end


Beyond/beyond_frame.lua
-- Globals
cur_xpos_g = 0
cur_ypos_g = 0
fr_width_g = 75
fr_height_g = 20

-- Set to 1 for debug messages; Or use /beyond_debug
beyond_debug = 0

-- Frame names & Hex color array
beyond_f_array = {"IT", "IA", "IPT", "CT", "CA", "CPT", "AT", "AA", "APT", "PT", "MA", "PA", "MP", "PPT"}
beyond_c_array = {
    "000011",
    "000022",
    "000033",
    "000044",
    "000055",
    "000066",
    "000077",
    "000088",
    "000099",
    "0000AA",
    "0000BB",
    "0000BB",
    "0000CC",
    "0000DD",
    "0000EE"
}

-- WowAce3
Beyond = LibStub("AceAddon-3.0"):NewAddon("Beyond", "AceConsole-3.0")

-- OnInitialize Event called by Ace3
function Beyond:OnInitialize()
    --self:Debug("Beyond:OnInitialize()")
    self:AddSlashCommands()
    self:Setup()
end

function Beyond:AddSlashCommands()
    -- Command prefix to listen for
    self:RegisterChatCommand("beyond", "HandleSlashCommand")
    self:RegisterChatCommand("br", "HandleSlashCommand")
end

function Beyond:HandleSlashCommand(cmd)
    -- Check command and run
    cmd = cmd and cmd:lower() or ""
    if cmd == "init" then
        self:FInit()
    elseif cmd == "debug" then
        self:Debug_toggle()
    elseif cmd == "toggle" then
        self:FToggle()
    else
        return
    end
end

function Beyond:Cords()
    -- Creating next cords for frame
    local curx
    local cury
    curx = cur_xpos_g
    cury = cur_ypos_g

    cur_xpos_g = cur_xpos_g + fr_width_g
    --self:Debug("Next xpos: " .. curx .. " ypos: " .. cury)
    return curx, cury
end

function Beyond:Setup()
    -- Setup frames for use
    local curx
    local cury
    local curw
    local curh
    curw = fr_width_g
    curh = fr_height_g

    for i, v in pairs(beyond_f_array) do
        curx, cury = self:Cords()
        self:Fcreate(beyond_f_array[i], curx, cury, curw, curh)
        --self:Debug("Create frame: " .. beyond_f_array[i] .. " xpos: " .. curx .. " ypos: " .. cury)
    end
end

function Beyond:FInit()
    -- Using 2 arrays to recolor all frames to test functionality
    for i, v in pairs(beyond_f_array) do
        curx, cury = self:Cords()
        --self:Debug("Change: " .. beyond_f_array[i] .. " Hex: " .. beyond_c_array[i])
        self:FSetHex(beyond_f_array[i], beyond_c_array[i])
    end
end

function Beyond:Debug(msg)
    --Display debug message if debug is true
    if beyond_debug == 1 then
        -- Display message if debug = true
        DEFAULT_CHAT_FRAME:AddMessage(msg)
        UIErrorsFrame:AddMessage(msg, 1.0, 1.0, 0, 1, 10)
    else
        -- Don't display message if debug is false
        return
    end
end

function Beyond:Debug_toggle()
    --Used for changing debug status on the fly
    if beyond_debug == 0 then
        beyond_debug = 1
        self:Debug("Debug ON")
    else
        self:Debug("Debug OFF")
        beyond_debug = 0
    end
end

function Beyond:Ftoggle(fname)
    local full_name = "btp_frame_" .. fname
    local frame = getglobal(full_name)
    --self:debug(full_name)
    if (frame) then
        if (frame:IsVisible()) then
            frame:Hide()
        else
            frame:Show()
        end
    end
end

function Beyond:Fcreate(arg_fname, curx, cury, curw, curh)
    local fname = "btp_frame_" .. arg_fname
    -- self:Debug("btp_frame_create: " .. fname .. " curx: " .. curx .. " cury: " .. cury .. " curw: " .. curw .. " curh: " .. curh)

    local f = CreateFrame("FRAME", fname, UIParent)
    f:SetWidth(curw)
    f:SetHeight(curh)
    f:SetPoint("TOPLEFT", UIParent, "TOPLEFT", curx, cury)
    f:SetBackdrop(
        {
            bgFile = "Interface\\Addons\\btp\\white16x16",
            tile = true,
            tileSize = 16,
            edgeFile = "Interface\\Addons\\btp\\white16x16",
            edgeSize = 1,
            insets = {left = 1, right = 1, top = 1, bottom = 1}
        }
    )
    f:SetBackdropBorderColor(0, 0, 0, 0)
    f:SetBackdropColor(0, 0, 0, 1)

    local fs = f:CreateFontString("$parentText", "ARTWORK", "GameFontNormal")
    fs:SetAllPoints()
    --fs:SetText(arg_fname)
end

function Beyond:FSetColor(fname, red, green, blue)
    local full_name = "btp_frame_" .. fname
    --self:Debug("btp_frame_set_color: " .. full_name ..
    --                " = " .. red .. " " .. " " .. green .. " " .. blue)
    local frame = getglobal(full_name)
    if (frame and red and green and blue) then
        frame:SetBackdropColor(red, green, blue)
    end
end

function Beyond:FSetHex(fname, hex)
    if (fname and hex) then
        local rhex, ghex, bhex = string.sub(hex, 1, 2), string.sub(hex, 3, 4), string.sub(hex, 5, 6)
        self:FSetColor(fname, tonumber(rhex, 16) / 255, tonumber(ghex, 16) / 255, tonumber(bhex, 16) / 255)
    end
end
